packages/hardhat-core/test/internal/hardhat-network/provider/node.ts

```diff
@@ -730,6 +757,188 @@ describe("HardhatNode", () => {
+        value: new BN(0),
+        data: toBuffer(contractInterface.encodeFunctionData("Hello", [])),
+        gasLimit: new BN(1_000_000),
+      };
+
+      function decodeResult(runCallResult: RunCallResult) {
+        return contractInterface.decodeFunctionResult(
+          "Hello",
+          bufferToHex(runCallResult.result.value)
+        )[0];
+      }
+
+      return decodeResult(
+        await targetNode.runCall({ ...callOpts, ...gasParams }, new BN(block))
+      );
+    }
+
+    describe("when forking with a default hardfork activation history", function () {
+      let hardhatNode: HardhatNode;
+
+      before(async function () {
+        [, hardhatNode] = await HardhatNode.create(baseNodeConfig);
+      });
+
+      it("should accept post-EIP-1559 gas semantics when running in the context of a post-EIP-1559 block", async function () {
+        assert.equal(
+          "Hello World",
+          await runCall(post1559GasOpts, post1559Block, hardhatNode)
+        );
+      });
+
+      it("should accept pre-EIP-1559 gas semantics when running in the context of a pre-EIP-1559 block", async function () {
+        assert.equal(
+          "Hello World",
+          await runCall(pre1559GasOpts, blockBefore1559, hardhatNode)
+        );
+      });
+
+      it("should throw when given post-EIP-1559 gas semantics and when running in the context of a pre-EIP-1559 block", async function () {
+        await expectErrorAsync(async () => {
+          assert.equal(
+            "Hello World",
+            await runCall(post1559GasOpts, blockBefore1559, hardhatNode)
+          );
+        }, "Cannot run transaction: EIP 1559 is not activated.");
+      });
+
+      it("should accept pre-EIP-1559 gas semantics when running in the context of a post-EIP-1559 block", async function () {
+        assert.equal(
+          "Hello World",
+          await runCall(pre1559GasOpts, post1559Block, hardhatNode)
+        );
+      });
+    });
+
+    describe("when forking with a hardfork activation history that indicates London happened one block early", function () {
+      let nodeWithEarlyLondon: HardhatNode;
+
+      before(async function () {
+        const nodeConfig = {
+          ...baseNodeConfig,
+          chains: cloneChainsConfig(baseNodeConfig.chains),
+        };
+
+        const chainConfig = nodeConfig.chains.get(1) ?? {
+          hardforkHistory: new Map(),
+        };
+        chainConfig.hardforkHistory.set(
+          HardforkName.LONDON,
+          eip1559ActivationBlock - 1
+        );
+
+        nodeConfig.chains.set(1, chainConfig);
+
+        [, nodeWithEarlyLondon] = await HardhatNode.create(nodeConfig);
+      });
+
+      it("should accept post-EIP-1559 gas semantics when running in the context of the block of the EIP-1559 activation", async function () {
+        assert.equal(
+          "Hello World",
+          await runCall(post1559GasOpts, blockBefore1559, nodeWithEarlyLondon)
```

Comment:
These tests are much, much clearer now, thanks.

This one is a bit surprising but I think this behavior is fine (and even if it isn't, I don't think we should put a lot of thought into what happens if someone uses the wrong hardfork block numbers).